### PR TITLE
smt: use existing bitWidth API

### PR DESCRIPTION
**Pull request by edwardcwang**
#### Type of Improvement

refactoring/documentation

#### API Impact
no impact
#### Backend Code Generation Impact
should have no impact
#### Desired Merge Strategy
squash
#### Release Notes
Refactor SMT backend to use the existing public API `firrtl.bitWidth`.
Also adds scaladoc to `firrtl.bitWidth`.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (1.2.x, 1.3.0, 1.4.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

File: src/main/scala/firrtl/Utils.scala [hunk 1]
```diff
@@ -51,7 +51,17 @@ object getWidth {
   def apply(e: Expression): Width = apply(e.tpe)
 }
 
+/**
+ * Helper object for computing the width of a firrtl type.
+ */
 object bitWidth {
```
Comment (ekiwi) [inline src/main/scala/firrtl/Utils.scala:57]: This is great! Thanks for discovering this function. I did not know that it existed.

File: src/main/scala/firrtl/backends/experimental/smt/Btor2Serializer.scala [hunk 2]
```diff
@@ -48,15 +48,15 @@ private class Btor2Serializer private () {
...
 
-  private def t(width: Int): Int = bitVecTypeCache.getOrElseUpdate(width, line(s"sort bitvec $width"))
+  private def t(width: BigInt): Int = bitVecTypeCache.getOrElseUpdate(width, line(s"sort bitvec $width"))
 
   // bit vector expression serialization
   private def s(expr: BVExpr): Int = expr match {
     case BVLiteral(value, width) => lit(value, width)
     case BVSymbol(name, _)       => symbols(name)
-    case BVExtend(e, 0, _)       => s(e)
+    case BVExtend(e, by, _) if by == 0 => s(e)
```
Comment (ekiwi) [inline src/main/scala/firrtl/backends/experimental/smt/Btor2Serializer.scala:59]: This is why I did not want to use BigInt: They do not work well for pattern matching.

File: src/main/scala/firrtl/backends/experimental/smt/FirrtlExpressionSemantics.scala [hunk 3]
```diff
@@ -8,19 +8,10 @@ import firrtl.PrimOps
 import firrtl.passes.CheckWidths.WidthTooBig
 
 private trait TranslationContext {
-  def getReference(name: String, tpe: ir.Type): BVExpr = BVSymbol(name, FirrtlExpressionSemantics.getWidth(tpe))
+  def getReference(name: String, tpe: ir.Type): BVExpr = BVSymbol(name, firrtl.bitWidth(tpe))
```
Comment (ekiwi) [inline src/main/scala/firrtl/backends/experimental/smt/FirrtlExpressionSemantics.scala:11]: This is good

**Review by ekiwi — CHANGES_REQUESTED** [review]
Replacing the custom `getWidth` function with one from the central utilities file is great. I did not know about it when writing the backend.

I am skeptical about the benefit of using BigInt for bit vector widths since they make pattern matching harder for no apparent benefit. The SMT Expression library has a type system that is completely separate from the firrtl type system (for one we do not distinguish SInt and UInt), and thus it does not necessarily have to follow the firrtl decision to use BigInt for widths.

Of course, if you can show me a use case where you even get close to the limit of an `Int` I am happy to make the change to BigInt.

File: src/main/scala/firrtl/backends/experimental/smt/Btor2Serializer.scala (hunk 2 above)
Comment (edwardcwang) [reply src/main/scala/firrtl/backends/experimental/smt/Btor2Serializer.scala]: No worries, that's fine - will put it back to Int. I don't have a strong opinion either way. Was just noting that for some reason the rest of FIRRTL uses BigInt for widths.

File: src/main/scala/firrtl/Utils.scala (hunk 1 above)
Comment (edwardcwang) [reply src/main/scala/firrtl/Utils.scala:57]: Real credit goes to @jackkoenig  :)

**Review by ekiwi — APPROVED** [review]
LGTM, you just need to run `sbt scalafmtAll` to make the CI happy.

**Comment by edwardcwang** [comment]
(useful note, `mill firrtl[2.12.13].reformat` also seems to do reformatting)

**Pull request merged**